### PR TITLE
Migrate to webpack@4

### PR DESCRIPTION
**Pull request by aulisius**
Hey,

Thanks for this plugin. Really useful =). I made the appropriate changes for this plugin to work w/o warnings with webpack@4

### CODE REVIEW

**Comment by aulisius** [comment]
The tests fail on node <=0.12. Are these a valid target to support? Even node@4 is going to be deprecated soon.

File: index.js [hunk 1]
```diff
@@ -178,21 +178,33 @@ var clean = function() {
...
     if (_this.options.watch) {
-      compiler.plugin("compile", function(params) {
+      const compile = function (params) {
         clean.call(_this);
-      });
+      }
+      if (compiler.hooks) {
+        compiler.hooks.compile.tap(pluginName, compile);
+      } else {
+        compiler.plugin("compile", compile);
```
Comment (yordis) [inline index.js]: I would just use `hooks`, if people want to stay on `<4` then do not upgrade the package.

Comment (aulisius) [reply index.js]: Cool. Will do those changes

Comment (aulisius) [reply index.js]: Can I use lambda and other ES6 stuff?

Comment (yordis) [reply index.js]: @aulisius dont forget Node 6, so as long as you support Node 6.

Also, there is a bunch of changes that are not part of the fix, could you remove those changes so the commit is cleaner.

File: index.js [hunk 2]
```diff
@@ -183,14 +183,15 @@ CleanWebpackPlugin.prototype.apply = function(compiler) {
   if (compiler === undefined) {
     return clean.call(_this);
   } else {
+    var pluginName = "clean-webpack-plugin";
```
Comment (yordis) [inline index.js]: Move this variable declaration to the top of the file please, this is something really unique and probably is better to extract it from the ejection.

**Comment by chrisblossom** [comment]
This PR removes support for Webpack v1-v3. [Here is an example](https://github.com/chrisblossom/clean-self-webpack-plugin/blob/e923f1c28c0b7903095f0035b17f068e203b960d/src/clean-self-webpack-plugin.js#L50-L77) of how to properly back support v1-v3 as well as v4.

**Comment by yordis** [comment]
@chrisblossom 😢 this was my fault, he added the support at first, @aulisius sorry.

**Comment by chrisblossom** [comment]
@yordis If this is expected, please disregard my comment. Although since Webpack 4 was just recently released and most people still using 2/3, I'd expect this plugin to work with at very least v3. @johnagan, thoughts?

**Comment by aulisius** [comment]
How are we going forward with this? @johnagan @yordis @chrisblossom  ?

**Comment by yordis** [comment]
@aulisius see the link that @chrisblossom gave you about supporting the legacy stuff, he has a good point. Just follow it (adding support for v3/v2) and we should good to go

**Comment by aulisius** [comment]
@yordis I believe we are good to go

**Comment by yordis** [comment]
@johnagan do you think we could have this merge as soon as possible please?

**Comment by chrisblossom** [comment]
This looks good to me. Thanks @aulisius!

**Comment by johnagan** [comment]
thanks all!

**Comment by johnagan** [comment]
`0.1.19`

**Pull request merged**